modules/core/client/services/menu.client.service.js

```diff
@@ -0,0 +1,196 @@
+
+      // Push new menu item
+      service.menus[menuId].items.push({
+        title: options.title || '',
+        state: options.state || '',
+        type: options.type || 'item',
+        class: options.class,
+        roles: ((options.roles === null || typeof options.roles === 'undefined') ? service.defaultRoles : options.roles),
+        position: options.position || 0,
+        items: [],
+        shouldRender: shouldRender
+      });
+
+      // Add submenu items
+      if (options.items) {
+        for (var i in options.items) {
+          if (options.items.hasOwnProperty(i)) {
+            service.addSubMenuItem(menuId, options.state, options.items[i]);
+          }
+        }
+      }
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    // Add submenu item object
+    function addSubMenuItem(menuId, parentItemState, options) {
+      options = options || {};
+
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Search for menu item
+      for (var itemIndex in service.menus[menuId].items) {
+        if (service.menus[menuId].items[itemIndex].state === parentItemState) {
+          // Push new submenu item
+          service.menus[menuId].items[itemIndex].items.push({
+            title: options.title || '',
+            state: options.state || '',
+            roles: ((options.roles === null || typeof options.roles === 'undefined') ? service.menus[menuId].items[itemIndex].roles : options.roles),
+            position: options.position || 0,
+            shouldRender: shouldRender
+          });
+        }
+      }
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    // Get the menu object by menu id
+    function getMenu(menuId) {
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Return the menu object
+      return service.menus[menuId];
+    }
+
+    function init() {
+      // A private function for rendering decision
+      shouldRender = function (user) {
+        if (!!~this.roles.indexOf('*')) {
+          return true;
+        } else {
+          if (!user) {
+            return false;
+          }
+
+          for (var userRoleIndex in user.roles) {
+            if (user.roles.hasOwnProperty(userRoleIndex)) {
+              for (var roleIndex in this.roles) {
+                if (this.roles.hasOwnProperty(roleIndex) && this.roles[roleIndex] === user.roles[userRoleIndex]) {
+                  return true;
+                }
+              }
+            }
+          }
+        }
+
+        return false;
+      };
+
+      // Adding the topbar menu
+      addMenu('topbar', {
+        roles: ['*']
+      });
+    }
+
+    // Remove existing menu object by menu id
+    function removeMenu(menuId) {
+      // Validate that the menu exists
+      service.validateMenuExistance(menuId);
+
+      // Return the menu object
+      delete service.menus[menuId];
```

Comment:
Actually this is fine, though it would probably be better if we did `service.menus[menuId] = undefined` because most browsers can more easily optimize that than an object that changes. (https://www.smashingmagazine.com/2012/11/writing-fast-memory-efficient-javascript/) That's a bit old, but written by a software engineer at Google who works on Chrome, so it's probably pretty accurate for most JavaScript engines.